pkg/loki/modules.go

```diff
@@ -350,6 +355,72 @@ func (t *Loki) initQueryFrontend() (_ services.Service, err error) {
 	}), nil
 }
 
+func (t *Loki) initRulerStorage() (_ services.Service, err error) {
+	// if the ruler is not configured and we're in single binary then let's just log an error and continue.
+	// unfortunately there is no way to generate a "default" config and compare default against actual
+	// to determine if it's unconfigured.  the following check, however, correctly tests this.
+	// Single binary integration tests will break if this ever drifts
+	if t.cfg.Target == All && t.cfg.Ruler.StoreConfig.IsDefaults() {
+		level.Info(util.Logger).Log("msg", "RulerStorage is not configured in single binary mode and will not be started.")
```

Comment:
Why no storage for single binary?

Reply:
This is just copied from Cortex, but it's basically just saying to treat the default (think zero value) ruler storage config as unconfigured when in the single binary.

Reply:
nevermind, I misunderstood the comment, this is basically saying don't bother starting the ruler if it's not configured.